### ISSUE TITLE
Active Tiles runs macro as GM not Player

### ISSUE THREAD

**Issue by trdischat** [ironmonk88/monks-active-tiles#11]
I am trying to create a tile that will run a macro as the player when the player steps on the tile.  This works with Trigger Happy and Advanced Macros.  But with Active Tiles the macro is run as the GM, even though the macro is configured to NOT run as GM.  This is important, because the macro needs to identify the player who stepped on the tile using `game.userId`.  When Active Tiles runs this code, it returns the userId of the GM, not the player.  Running the exact same macro with Trigger Happy and Advanced Macros returns the userId of the player.

**Comment by mjeffw** [comment]
I'll second this request. I have any number of macros that apply effects to the currently active/selected token. If the macro can be configured to run as the player of the triggering token, these macros would all work. As it is now, I have to (as the GM) quickly select the player's token before they move in case they step on an Active Tile or else the macro applies to whatever other token I have selected.

This will make the difference between using Active Tiles in my game, or not. (No pressure, man!) ;-)

**Comment by ironmonk88** [comment]
For various reasons the GM needs to run all the actions.
The latest version allows the GM to pass the macro execution back to the player.

**Comment by trdischat** [comment]
Not working yet. It looks like my macro is grabbing the right userId. But I now get this error message: "There was an error in your macro syntax. See the console (F12) for details."  The console does not provide details that really explain the syntax problem.  Stepping on the tile is supposed to run this macro as the Player with a text string like "NameOfBreakoutRoom" as the first parameter.
```js
// Collect user and breakout names, then launch breakout
// Run this macro NOT as GM
const userId = args[1] || game.userId;           // id of the user to send to breakout
let breakoutRoom = args[0];                      // breakout room name
if (breakoutRoom) {                              // if room name blank, exit breakout
    const rmStr = args[0].hashCode();                // hash the breakout room name
    const gmStr = game.world.data.title.hashCode();  // hash the world title
    let x = new Date();
    x.setTime(x.valueOf() + ( 60000 * x.getTimezoneOffset() ));
    const dtStr = x.toDateString().hashCode();       // hash the current date in GMT 
    breakoutRoom = rmStr + gmStr + dtStr;      // use combined hashes as room name
}
let RunBreakOut = game.macros.getName("RunBreakOut");
RunBreakOut.execute(userId, breakoutRoom);
```
The first macro then calls the second macro, which runs as the GM.
```js
// Launch Jitsi breakout room
// Run this macro as GM
const userId = args[0];
const breakoutRoom = args[1];
game.webrtc.client._jitsiClient.settings.set("client", `users.${userId}.jitsiBreakoutRoom`, breakoutRoom);
game.socket.emit("module.jitsirtc", { action: "breakout", userId, breakoutRoom });
```
Using the Advanced Macros module, the Player can type `/BreakOut SendMeAway` in Chat and be sent into a separate Jitsi session.  Using Trigger Happy, the Player gets sent to a separate Jitsi session as soon as they step on the trigger object.  So I know that the macros work as intended.  Active Tiles throws the syntax error and only shows in the console an object with the name of the breakout room (which is the value of args[0]), and keys for actor, tile, token, and userid.

**Comment by mjeffw** [comment]
Can I get the ID of the token that triggered the Active Tile in a macro?

**Comment by ironmonk88** [comment]
If you put `debugger;` at the top of your macro, it will launch the debug console when it reaches that code.  And then you can step through the code and see what's going wrong.  I'm more than likely not passing in the arguments properly.

**Comment by trdischat** [comment]
When invoking a macro directly or with Trigger Happy, args[0] is the first parameter passed from the Advanced Macros chat command.  In the case of my macro, args[0] is the name of the breakoutRoom.  Active Tiles overrides this and passes an object in the form of `{actor: Actor5e, tile: TileDocument, token: TokenDocument5e, userid: String}` as args[0], pushing the real first parameter down to args[1]. This breaks whatever the macro is expecting in the way of parameters.

**Comment by ironmonk88** [comment]
Unfortunately it doesn't look like I can change that due to the way Active Tiles does things.  If the Macro is set to "Run as GM" Advanced Macros will change the token/actor/speaker to the current Token that the GM has control of rather than the Token that triggered the tile.  

So the macro will run against the GMs character instead of the players.  Adding the object out front allows Advanced Macros to use that information instead, so the correct Token is reported to the macro.  I've changed it so that args[0] contains a property called args.  Which is where the data goes.  So your macros would have to change to args[0].args[0] unfortunately to be used properly by Active Tiles.

I tried my best to match Trigger Happy, but it looks like it'll have to be different for the moment.

That being said... if Advanced Macros was updated to look through all arguments instead of just the first one, it could better match what Trigger Happy does.
